[distributed][rpc] separate out rpc to sync and async apis

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26570 [distributed][rpc] separate out rpc to sync and async apis**

Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

rohan-varma committed:
[pytorch][distributed][rpc] separate out rpc to sync and async apis

This diff splits out the async and sync rpc implementations to their
own functions. This way, we can have stronger typehinting and make the type of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the functiohn.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -158,6 +159,105 @@ def remote(to, func, args=None, kwargs=None):
+
+
 @_require_initialized
 def rpc(to, func, args=None, kwargs=None, async_call=False):
```
Hey @rohan-varma, thanks for adding this. IIUC, the original goal was actually removing `rpc` and splitting it into `rpc_sync` and `rpc_async`. The reason for this is that, if `rpc` does not have a fixed return type, we won't be able to TorchScript it.

rohan-varma replied:
@mrshenli Are we concerned about backwards compatibility here? I'm not completely familiar with the backwards compatibility protocols we follow, but won't completely removing `rpc` immediately break the code of whoever is using it when they update to the new version?

mrshenli replied:
Currently `test/test_rpc.py` and `test/test_dist_autograd.py` are the only dependency we have I believe, it should be sufficient to modify those accordingly. But I think your solution is more appropriate by marking `rpc` as deprecated. Let's land this PR and then remove `dist.rpc` later. :)

mrshenli reviewed (approved):
LGTM!

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -215,28 +315,17 @@ def rpc(to, func, args=None, kwargs=None, async_call=False):
-            _agent, _to_worker_id(to), serialize(PythonUDF(func, args, kwargs)))
-
+    warnings.warn(
+        """dist.rpc is deprecated. Use dist.rpc_async for asynchronous
```
nit: shall we modify `test_rpc.py` and `test_dist_autograd` in this PR as well? Otherwise, this warning message will show up in tests?

rohan-varma replied:
Yep, sounds good

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed][rpc] separate out rpc to sync and async apis"


Per #24247, this splits out the async and sync rpc implementations to their
own functions. Previously, the user would call `dist.rpc` and pass in an `async` flag if they wanted to run it asynchronously. This change introduces two new methods: `rpc_sync` and `rpc_async` that return the result or a future containing the result, respectively. The common code is moved to `_invoke_rpc`.

This way, we can have stronger type hinting and make the type (async vs sync) of
RPC being done more explicit to our users, as opposed to them having to pass
in an async flag to the function.

Differential Revision: [D17509975](https://our.internmc.facebook.com/intern/diff/D17509975/)

[ghstack-poisoned]

peterjc123 commented on test/test_rpc.py:
```diff
@@ -3,21 +3,22 @@
 import torch.distributed as dist
+from common_distributed import MultiProcessTestCase
+from common_utils import load_tests, run_tests
+from torch.distributed.rpc import RpcBackend
```
This section should be put after the if check, otherwise the import will fail. This is currently blocking the Windows builds.

rohan-varma replied:
@peterjc123 thanks for pointing this out, I am putting up a diff now.

rohan-varma replied:
@peterjc123 https://github.com/pytorch/pytorch/pull/27126/files

rohan-varma replied:
Fixed in 27126.

rohan-varma commented:
Windows build issue fixed in #27126.
